Final charm backend

epmikida opened the pull request:
Here is the PR for merging in the Charm++ backend to FleCSI.

To build and test the Charm++ backend:

1. Clone Charm++ from https://github.com/UIUC-PPL/charm
2. Build Charm++ by running `./build charm++ mpi-linux-x86_64 --build-shared` from within the top level charm directory
3. Build all other FleCSI dependencies as if you were building the Legion backend (except that Legion is not needed)
4. Change the following cmake variables within FleCSI

FLECSI_RUNTIME_MODEL	charm
CMAKE_CXX_COMPILER	<path_to_charm/bin/charmc>
CMAKE_CXX_FLAGS		-c++-option --std=gnu++17
CMAKE_C_COMPILER		<path_to_charm/bin/charmc>
CMAKE_C_FLAGS			-c++-option --std=gnu++17

epmikida committed:
Create a build skeleton for the Charm++ backend

epmikida committed:
Change CMake to allow for .ci compilation

epmikida committed:
Create charm backend from copy of legion backend

epmikida committed:
Fix CMAKE for charm backend

epmikida committed:
Update context to use charm runtime instead of legion

epmikida committed:
Remove rest of legion code from context

epmikida committed:
Get tasks executing inline with non-flecsi data

epmikida committed:
Quick changes to get global tests passing with basic field reg

epmikida committed:
Clean out some old legion code from exec

epmikida committed:
More aggressively remove legion from charm build

epmikida committed:
Add dummy future and make test tasks function properly

epmikida committed:
Fix test build dependency

epmikida committed:
Add more to field registration code

epmikida committed:
Remove decl and def files from repo

epmikida committed:
Bring up to date with devel API changes

epmikida committed:
Minor cleanup and commenting

epmikida committed:
Small fixes to wrap up recent rebase

opensdh commented on flecsi/data/backend.hh:
```diff
@@ -46,4 +46,8 @@ constexpr inline std::size_t logical_size = 1ul << 32;
 
+#elif FLECSI_RUNTIME_MODEL == FLECSI_RUNTIME_MODEL_charm
+
+#include <flecsi/data/charm/policy.hh>
```
We've switched to using the style `#include "flecsi/data/charm/policy.hh"` everywhere (though we haven't proactively updated everything, obviously).

opensdh commented on flecsi/exec/charm/bind_accessors.hh:
```diff
@@ -0,0 +1,112 @@
+  the Charm++ backend.
+ */
+
+struct bind_accessors_t : public util::tuple_walker<bind_accessors_t> {
```
Let's please not add usage of `util::tuple_walker`; now that we use C++17, it's very easy to provide something like
```cpp
template<class T>
void operator()(T &&tup) {
  std::apply([this](TT &... tt) { (visit(tt), ...); }, std::forward<T>(tup));
}
```
Also, new classes shouldn't use `_t` (which signifies POSIX types and trait aliases).

opensdh commented on flecsi/exec/charm/bind_accessors.hh:
```diff
@@ -0,0 +1,112 @@
+  */
+  template<typename DATA_TYPE, size_t PRIVILEGES>
+  void visit(data::accessor<data::dense, DATA_TYPE, PRIVILEGES> & accessor) {
+    flog_assert(buf_.size() % sizeof(DATA_TYPE) == 0, "Bad buffer size\n");
```
This is the one use of `buf_`, and (as per the comment above) it doesn't seem appropriate (what if there were more than one accessor in a call?).

opensdh commented on flecsi/exec/charm/policy.hh:
```diff
@@ -0,0 +1,195 @@
+// is_constructible_v<float&,double&> instead.
+template<class... PP, class... AA>
+auto
+serial_arguments(std::tuple<PP...> * /* to deduce PP */, AA &&... aa) {
```
If all this machinery (including all the serialization in `task_wrapper.hh`) is needed for multiple backends, it should be factored out somewhere rather than duplicated.

opensdh commented on flecsi/exec/charm/policy.hh:
```diff
@@ -0,0 +1,195 @@
+template<class, class>
+struct tuple_prepend;
+template<class T, class... TT>
+struct tuple_prepend<T, std::tuple<TT...>> {
```
Fair warning: this is probably going to go away (it turned out that having "MPI tasks" have an assumed number of point tasks was unnecessarily restrictive).  (Sorry that `devel` is such a moving target in places.)

opensdh commented on flecsi/exec/charm/task_prologue.hh:
```diff
@@ -0,0 +1,169 @@
+      field_reference<DATA_TYPE, data::raw, topo::global, topo::elements> &
+        ref) {
+    auto & flecsi_context = run::context::instance();
+    flecsi_context.regField(ref.fid(), sizeof(DATA_TYPE));
```
Why should we be registering a field for every use of it for a task launch?

opensdh commented on flecsi/exec/charm/task_wrapper.hh:
```diff
@@ -0,0 +1,254 @@
+
+ */
+
+template<auto & F, task_processor_type_t P> // P is for specialization only
```
Evidently this comment was a lie in the version copied (it's used 7 lines later).  (Sorry about that!)

opensdh commented on flecsi/run/charm/context.cc:
```diff
@@ -0,0 +1,123 @@
+
+  if(context::initialize_dependent_) {
+    CharmLibExit();
+    MPI_Finalize();
```
Does `CharmBeginInit` call `MPI_Init`, so that this balances that?

epmikida replied:
This actually needs to change. We made some small changes to the Charm interop interfaces while working with Flecsi. CharmInit calls MPI_Init, and CharmLibExit() calls MPI_Finalize, so I'll need to remove the extra call to MPI_Finalize() here.

opensdh replied:
I thought it might, given the `argc`/`argv` arguments.  Without claiming to know anything about Charm++, I wonder why you want that: obviously any two libraries that do so can't be used together.

epmikida replied:
It depends on the mode you are using Charm++ in and the type of Charm++ build you are doing. The mode we have chosen for Flecsi is what we refer to as user-driven mode. The user calls CharmInit and CharmLibExit in basically the same way they would call MPI_Init/MPI_Finalize. Those calls initialize and finalize the underlying communication layers. If you build the MPI version of Charm++, then that would include initializing and finalizing MPI. However, if Flecsi doesn't have an inherent need for MPI, then you could build Charm++ on top of any communication layer you want.

If instead you are planning on explicit interoperation with Charm++ and MPI, you would then use the interop mode for Charm++. In that mode, you are responsible for initializing and finalizing MPI yourself, and when you initialize Charm++ you give it the MPI communicator you want it to live within. In this mode you can initialize as many charm instances as you want, and CharmLibExit() has no effect on your MPI environment.

Since Flecsi doesn't explicitly rely on MPI to our knowledge, we figured the user driven mode made the most sense. This extra MPI_Finalize should have been removed for sure, but I think I just missed it when preparing the pull request.

opensdh replied:
Good to know.  While we don't assume that FleCSI clients will be absolutely portable between backends, it is worth noting that we expect clients that target Legion to occasionally use MPI directly (in so-called "MPI tasks": `flecsi::execute<foo,flecsi::mpi>(…)`).  Whether that means that they will care about initializing Charm++ themselves or using something other than `MPI_COMM_WORLD` for it, I can't say.

epmikida replied:
Ah OK. That's certainly something the Charm backend could also support pretty easily. The initialize and finalize methods would then just call MPI_Init/MPI_Finalize like they do in Legion, as well as calling our interop CharmInit and CharmExit. Part of the reason we didn't do that initially is because that would restrict us to using a Charm that is built on top of MPI. Whereas in the current version, you can use Charm++ built on whatever communication layer makes the most sense for where you are running it.

opensdh commented on flecsi/run/charm/context.hh:
```diff
@@ -0,0 +1,258 @@
+namespace flecsi::run {
+
+const size_t FLECSI_TOP_LEVEL_TASK_ID = 0;
+const size_t FLECSI_MAPPER_FORCE_RANK_MATCH = 0x00001000;
```
Are these constants meaningful for this backend?

opensdh commented on flecsi/run/charm/context.hh:
```diff
@@ -0,0 +1,258 @@
+  auto execute(std::vector<std::byte>& buf) {
+    depth++;
+    return T::execute(buf);
+    depth--;
```
This line is unreachable.

opensdh commented on flecsi/exec/charm/bind_accessors.hh:
```diff
@@ -0,0 +1,112 @@
+  void visit(data::accessor<data::dense, DATA_TYPE, PRIVILEGES> & accessor) {
+    flog_assert(buf_.size() % sizeof(DATA_TYPE) == 0, "Bad buffer size\n");
+    auto & flecsi_context = run::context::instance();
+    DATA_TYPE* d = (DATA_TYPE*)flecsi_context.getField(accessor.identifier());
```
Please use `reinterpret_cast` to make it obvious what's happening here.

opensdh commented on flecsi/run/charm/context.hh:
```diff
@@ -0,0 +1,258 @@
+  /// Store a reference to the argument under a small unused positive integer.
+  /// Its type is forgotten.
+  template<class T>
+  std::size_t record(T & t) {
```
Is this facility needed?  (Legion doesn't use it anymore.)

opensdh reviewed (commented):
I did one pass; there may be more to say after certain simplifications.

epmikida committed:
Remove extra call to MPI_Finalize

CharmLibExit calls MPI_Finalize if Charm++ is built on the MPI layer, so this was causing crashes. If Charm is not built on the MPI layer, then MPI is never used/initialized in the first place.

epmikida committed:
Cleanup for merge

ollielo commented:
Closed due to lack of (discontinued) funding for this line of work.
